--- old_linux/linux/net/core/dev.c	2020-05-13 15:29:21.442897763 +0530
+++ linux/net/core/dev.c	2020-05-13 15:31:25.151376093 +0530
@@ -1787,9 +1787,9 @@
 
 			if (skb_network_header(skb2) < skb2->data ||
 			    skb_network_header(skb2) > skb_tail_pointer(skb2)) {
-				net_crit_ratelimited("protocol %04x is buggy, dev %s\n",
-						     ntohs(skb2->protocol),
-						     dev->name);
+				//net_crit_ratelimited("protocol %04x is buggy, dev %s\n",
+					//	     ntohs(skb2->protocol),
+						//     dev->name);
 				skb_reset_network_header(skb2);
 			}
 
